allennlp/data/dataset_readers/nlvr2.py

```diff
@@ -29,146 +47,120 @@ def __init__(
-        label: bool,
+        self, identifier: str, sentence: str, label: bool,  # type: ignore
     ) -> Instance:
-        if self.mask_prepositions_verbs:
-            doc = self.spacy(sentence)
-            prep_verb_starts = [
-                (token.idx, len(token))
-                for token in doc
-                if token.dep_ == "prep" or token.pos_ == "VERB"
-            ]
-            new_sentence = ""
-            prev_end = 0
-            for (idx, length) in prep_verb_starts:
-                new_sentence += sentence[prev_end:idx] + self._tokenizer.tokenizer.mask_token
-                prev_end = idx + length
-            new_sentence += sentence[prev_end:]
-            sentence = new_sentence
-        elif self.drop_prepositions_verbs:
-            doc = self.spacy(sentence)
-            prep_verb_starts = [
-                (token.idx, len(token))
-                for token in doc
-                if token.dep_ == "prep" or token.pos_ == "VERB"
-            ]
-            new_sentence = ""
-            prev_end = 0
-            for (idx, length) in prep_verb_starts:
-                new_sentence += sentence[prev_end:idx]
-                prev_end = idx + length
-            new_sentence += sentence[prev_end:]
-            sentence = new_sentence
         tokenized_sentence = self._tokenizer.tokenize(sentence)
-        from allennlp.data.fields import TextField
 
         sentence_field = TextField(tokenized_sentence, self._token_indexers)
 
         # Load images
         image_name_base = identifier[: identifier.rindex("-")]
-        images = [self.images[f"{image_name_base}-img{image_id}.png"] for image_id in [0, 1]]
-        images, sizes = self.image_loader(images)
-        featurized_images = self.image_featurizer(images, sizes)
-        
-        import torch
-        with torch.no_grad():
-            # I'm not happy about the squeezing and unsqueezing here.
-            proposals = self.proposal_generator(images, sizes, featurized_images)
-            # proposals = [self.proposal_generator(i.unsqueeze(0)).squeeze(0) for i in images]
-            visual_features = [
-                self.proposal_embedder(i.unsqueeze(0), p.unsqueeze(0)).squeeze(0)
-                for i, p in zip(images, proposals)
-            ]
-        from allennlp.data.fields import MetadataField
-
-        from allennlp.data.fields import ArrayField
-        from allennlp.data.fields import ListField
+
+        # TODO(mattg): shouldn't we be using the URL?  What if images are reused?  
+        # Since we use the image path, it will not be reused. 
+        left_image_path = self.images[f"{image_name_base}-img0.png"]
+        right_image_path = self.images[f"{image_name_base}-img1.png"]
+
+        to_compute = []
+        if left_image_path not in self._feature_cache:
+            to_compute.append(left_image_path)
+        if right_image_path not in self._feature_cache:
+            to_compute.append(right_image_path)
+
+        if to_compute:
```

Comment:
I just checked PEP8 (the standard python style guide) and it doesn't give an opinion on this point, though it does have a warning to be careful when doing this.  This is such a common usage in python that I expected to see something, and was surprised when it wasn't there.  Because it's so common, and because the list was created 6 lines above, I'm not too concerned with this particular case, but yes, I agree that avoid this except in really simple cases like this is probably a good idea.